File: src/python/problems/Subsets.py
#https://leetcode.com/problems/subsets/

from core.ProblemBase import *

class Subsets(ProblemBase):
    def Solution(self, nums: list[int]) -> list[list[int]]:
        result = []

        def search(index, temp):
            result.append(temp.copy())

            for i in range(index, len(nums)):
                temp.append(nums[i])
                search(i + 1, temp)
                temp.pop()

        search(0, [])

        return result

if __name__ == '__main__':
    TestGen(Subsets) \
        .Add(lambda tc: tc.Param("nums", [1,2,3]).Result([[],[1],[2],[1,2],[3],[1,3],[2,3],[1,2,3]])) \
        .Add(lambda tc: tc.Param("nums", []).Result([[],[0]])) \
        .Run()

File: src/csharp/Problems/Subsets.cs
//https://leetcode.com/problems/subsets/

namespace LeetCode.Problems;

public sealed class Subsets : ProblemBase
{
    [Theory]
    [ClassData(typeof(Subsets))]
    public override void Test(object[] data) => base.Test(data);

    public override void AddTestCases()
        => Add(it => it.ParamArray("[1,2,3]").Result2dArray("[[],[1],[1,2],[1,2,3],[1,3],[2],[2,3],[3]]", true))
          .Add(it => it.ParamArray("[0]").Result2dArray("[[],[0]]", true));

    private IList<IList<int>> Solution(int[] nums)
    {
        var result = new List<IList<int>>();

        Search(nums, 0, new List<int>(10), result);

        return result;
    }

    private void Search(int[] nums, int index, IList<int> temp, IList<IList<int>> result)
    {
        result.Add(temp.ToArray());
        
        for (int i = index; i < nums.Length; i++)
        {
            temp.Add(nums[i]);
            Search(nums, i + 1, temp, result);
            temp.RemoveAt(temp.Count - 1);
        }
    }
}